server/src-lib/Hasura/Metadata/Class.hs

```diff
@@ -32,38 +34,49 @@ runMetadataStorageT =
 
 {- Note [Todo: Common interface for eventing sub-system]
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-Scheduled and Table event triggers have their own implementation
-and tables in Postgres. The core logic for both is same. So it
-is necessary to have a unified interface (maybe via a Postgres extension)
-where all event triggers implementation can extend from. By this we can
-fairly reduce the Postgres schema foot print and interactions to the database.
+Postgres tables' event triggers and scheduled event triggers are similar in the
+core logic. But currently, their implementation is completely isolated and do not share
+a common schema in Postgres. We're having a plan to simplify them via a common interface
+(maybe via a Postgres extension). This will potentially reduce number of interactions made
+to database and schema foot print.
 
 TODO: Reference to open issue or rfc?
 -}
 
 -- | Metadata storage abstraction via a type class.
 --
+-- This type class enables storing and managing Hasura metadata in an isolated
+-- database which will not interfere with user's database where tables/functions
+-- are defined. Hence, it'll enable support for databases of multiple backends
+-- like MySQL, MSSQL etc.
```

Comment:
Why have these methods been abstracted in a type class, rather than just being implemented once?

Why do these three areas belong together in one type class? I think an element of the answer to this second question could be (if I understand correctly) that the "global state" of graphql-engine consists exactly of these three things: metadata, scheduled triggers, and async actions.

Reply:
> Why have these methods been abstracted in a type class, rather than just being implemented once?

Implementation will be different for OSS and Pro/Cloud graphql engines. Pro/Cloud graphql-engine code is using OSS code as library with there own implementation. 

> Why do these three areas belong together in one type class?

https://github.com/rakeshkky/graphql-engine/blob/664c43b1526d63cc12886089bcbc4897c353e91b/server/src-lib/Hasura/Metadata/Class.hs#L77